### DISCUSSION TITLE
Improve Nullability Analysis for Testing Frameworks

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/roslyn#54169] [Ideas]
I believe that Code such as:
```
Assert.IsNotNull(x);
Assert.AreNotEqual(x, null);
// etc.
```
Should cause nullability state to change.

**Comment by jmarolf** [comment]
@kendrahavens I assume we would want to move this to the respective test platforms?

**Reply by kendrahavens** [reply]
Sounds good to me. Could this be provided by vstest or do we need to open an issue on each test framework (xUnit, NUnit, MSTest)?

**Reply by ryzngard** [reply]
Test frameworks should be doing this I think.

**Reply by Haplois** [reply]
We added this support (MSTest v2) on https://github.com/microsoft/testfx/pull/744, and released it with v2.2.1.

![image](https://user-images.githubusercontent.com/433724/124444924-5fa27700-dd7f-11eb-9f73-9a08e0b4221f.png)

**Comment by MaceWindu** [comment]
```cs
T IsNotNull<T>(T? value) where T: struct {}
T IsNotNull<T>(T? value) where T: class {}


x = Assert.IsNotNull(x);
// x is not nullable here
```
